### ISSUE TITLE
Use pyproject.toml to control build-time numpy version

### ISSUE THREAD

**Issue by bmerry** [inducer/pycuda#271]
Currently pycuda uses a `setup_requires` in `setup.py`. That approach has a number of issues, as documented in PEP 518, and these days it's preferable to use a pyproject.toml to declare build-time dependencies that `pip` will install into an isolated environment.

An advantage of this is that it could pin the numpy version used at build time. Using the *oldest* supported numpy at build time will ensure that the resulting installation (and wheel) can be used with any newer numpy version. The reverse isn't true: if I do `pip install numpy==1.18.1 pycuda` in a fresh venv, the installation of pycuda may install the latest numpy, build against it, then pip will install the older numpy, and the result is errors like `RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd`.

Caveat: I don't know anything about non-pip environments like conda, nor how this would interact with aksetup. It's also a bit tricky to say what version should be specified, because ancient versions of numpy fail to build against recent versions of CPython.

**Comment by inducer** [comment]
Thanks for the suggestion!

I haven't used the `pyproject.toml` mechanism much, but I'm open to trying it. Would you be willing to put together a PR with a prototype?

An additional concern that popped into my head is that, at some point in the not-too-distant future, PyCUDA is hopefully going to make the jump from Boost.Python to pybind11. Pybind11 requires some [widgetry](https://github.com/pybind/pybind11/blob/master/setup.py) in `setup.py` for seamless installation. I don't know (but imagine) that this can in some way be made compatible with `pyproject.toml`.

**Comment by bmerry** [comment]
Sure, I can put together a draft PR some time.

I've used pybind11 with pyproject.toml before - it's pretty straightforward these days e.g. these three lines in one project I maintain:
- https://github.com/ska-sa/spead2/blob/409b33af66e642837346fee095f04eaf84ce1bdf/setup.py#L27
- https://github.com/ska-sa/spead2/blob/409b33af66e642837346fee095f04eaf84ce1bdf/setup.py#L133
- https://github.com/ska-sa/spead2/blob/409b33af66e642837346fee095f04eaf84ce1bdf/pyproject.toml#L2

New versions of pybind11 do have a tendency to break things so I've found it very useful to be able to specify the version in pyproject.toml.
